src/main/java/org/zalando/nakadi/service/subscription/SubscriptionTimeLagService.java

```diff
@@ -0,0 +1,129 @@
+
+@Component
+public class SubscriptionTimeLagService {
+
+    private static final int EVENT_FETCH_WAIT_TIME_MS = 1000;
+    private static final int COMPLETE_TIMEOUT_MS = 30000;
+    private static final int LAG_CALCULATION_PARALLELISM = 10;
+
+    private final TimelineService timelineService;
+    private final NakadiCursorComparator cursorComparator;
+
+    @Autowired
+    public SubscriptionTimeLagService(final TimelineService timelineService,
+                                      final NakadiCursorComparator cursorComparator) {
+        this.timelineService = timelineService;
+        this.cursorComparator = cursorComparator;
+    }
+
+    public Map<EventTypePartition, Duration> getTimeLags(final Collection<NakadiCursor> committedPositions,
+                                                         final List<PartitionEndStatistics> endPositions)
+            throws ErrorGettingCursorTimeLagException, InconsistentStateException {
+
+        final ExecutorService executor = Executors.newFixedThreadPool(LAG_CALCULATION_PARALLELISM);
+
+        final Map<EventTypePartition, Future<Duration>> futures = new HashMap<>();
+        final Map<EventTypePartition, Duration> result = new HashMap<>();
+
+        for (final NakadiCursor cursor : committedPositions) {
+            final EventTypePartition partition = new EventTypePartition(cursor.getTimeline().getEventType(),
+                    cursor.getPartition());
+            if (isCursorAtTail(cursor, endPositions)) {
+                result.put(partition, Duration.ZERO);
+            } else {
+                futures.put(partition, executor.submit(() -> getNextEventTimeLag(cursor)));
+            }
+        }
+
+        executor.shutdown();
+        try {
+            final boolean finished = executor.awaitTermination(COMPLETE_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+            if (!finished) {
+                throw new InconsistentStateException("Timeout occurred when getting subscription time lag");
+            }
+            for (final EventTypePartition partition : futures.keySet()) {
+                final Future<Duration> future = futures.get(partition);
+                try {
+                    result.put(partition, future.get());
+                } catch (final ExecutionException e) {
+                    throw e.getCause();
+                }
+            }
+            return result;
+
+        } catch (final InconsistentStateException | ErrorGettingCursorTimeLagException e) {
+            throw e;
+        } catch (final Throwable e) {
+            throw new InconsistentStateException("Unexpected error occurred when getting subscription time lag", e);
+        }
+    }
+
+    private Duration getNextEventTimeLag(final NakadiCursor cursor) throws ErrorGettingCursorTimeLagException,
+            InconsistentStateException {
+        try {
+            final EventConsumer consumer = timelineService.createEventConsumer(
```

Comment:
Creating a consumer is an expensive thing. Would you consider to try to "pack" many partitions in a single consumer? I'm not sure we have control over independent partitions reading, but I would like to ask you anyway, why did you decide not to do it.

Reply:
The problem here is exactly no control over reading from specific partitions. So if I create a consumer for many partitions I don't know when I will get data for some specific partition. It may happen that I will have to read a lot of data from partition A before I get at least something for partition B. 
That's why I decided to make a consumer per partition. And as that happens in multi-threaded way - I can't share the same consumer of course.

Reply:
Ok, got it.

Reply:
@v-stepanov I was checking Kafka Consumer parameters and maybe we can create a single consumer that will read a single event from each partition by using `max.partition.fetch.bytes = 1` http://kafka.apache.org/documentation.html#consumerconfigs

Reply:
Another thing is that it's not so nice to pass kafka-specific parameter from a place which is abstracted from kafka.